File: web/views/presets.js
export class PresetView {
    constructor(api) {
        this.api = api;
        this.element = null;
    }

    render() {
        this.element = document.createElement('div');
        this.element.className = 'preset-view';
        this.element.innerHTML = `
            <div class="preset-controls">
                <input type="text" id="preset-name" placeholder="Preset Name">
                <button id="save-btn">Save Current</button>
            </div>
            <div id="preset-list" class="preset-list">Loading...</div>
        `;

        this.element.querySelector('#save-btn').addEventListener('click', () => this.savePreset());
        this.loadPresets();

        return this.element;
    }

    async loadPresets() {
        const list = this.element.querySelector('#preset-list');
        try {
            const res = await fetch('/api/presets');
            const presets = await res.json();
            
            list.innerHTML = presets.map(name => `
                <div class="preset-item">
                    <span>${name}</span>
                    <button data-name="${name}" class="load-btn">Load</button>
                </div>
            `).join('');
            
            list.querySelectorAll('.load-btn').forEach(btn => {
                btn.addEventListener('click', () => this.loadPreset(btn.dataset.name));
            });
        } catch (e) {
            list.innerHTML = `Error: ${e.message}`;
        }
    }

    async savePreset() {
        const input = this.element.querySelector('#preset-name');
        const name = input.value.trim();
        if (!name) return;

        // Fetch current graph to save
        // Ideally backend does this, but for now we construct the preset object here?
        // Wait, API expects `Preset` struct.
        // We need to fetch current graph, convert links to LinkSpec, then send.
        
        try {
            const graph = await this.api.getGraph();
            const links = graph.links.map(l => {
                const outNode = graph.nodes.find(n => n.id === l.output_node);
                const inNode = graph.nodes.find(n => n.id === l.input_node);
                // Need ports too
                const outPort = outNode?.ports.find(p => p.id === l.output_port);
                const inPort = inNode?.ports.find(p => p.id === l.input_port);
                
                return {
                    output_node: outNode?.name || "",
                    output_port: outPort?.name || "",
                    input_node: inNode?.name || "",
                    input_port: inPort?.name || ""
                };
            }).filter(l => l.output_node && l.input_node);

            const preset = {
                name,
                description: "User saved preset",
                links
            };

            await fetch('/api/presets', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify(preset)
            });
            
            input.value = '';
            this.loadPresets();
        } catch (e) {
            console.error("Save failed", e);
        }
    }

    async loadPreset(name) {
        try {
            await fetch(`/api/presets/${name}/load`, { method: 'POST' });
            // Notify success
        } catch (e) {
            console.error("Load failed", e);
        }
    }
}

File: src/api/presets.rs
use axum::{
    extract::{State, Path},
    Json,
    http::StatusCode,
};
use crate::AppState;
use crate::graph::preset::Preset;

pub async fn list_presets(
    State(state): State<AppState>,
) -> Json<Vec<String>> {
    let list = state.preset_manager.list_presets().await.unwrap_or_default();
    Json(list)
}

pub async fn save_preset(
    State(state): State<AppState>,
    Json(preset): Json<Preset>,
) -> StatusCode {
    match state.preset_manager.save_preset(&preset).await {
        Ok(_) => StatusCode::OK,
        Err(e) => {
            eprintln!("Failed to save preset: {}", e);
            StatusCode::INTERNAL_SERVER_ERROR
        }
    }
}

pub async fn load_preset(
    State(state): State<AppState>,
    Path(name): Path<String>,
) -> StatusCode {
    let preset = match state.preset_manager.load_preset(&name).await {
        Ok(p) => p,
        Err(_) => return StatusCode::NOT_FOUND,
    };

    let graph = state.graph.read();
    let current_nodes = &graph.get_graph().nodes;

    for link_spec in preset.links {
        // Resolve Names to IDs
        let out_node = current_nodes.iter().find(|n| n.name == link_spec.output_node);
        let in_node = current_nodes.iter().find(|n| n.name == link_spec.input_node);

        if let (Some(out), Some(input)) = (out_node, in_node) {
            let out_port = out.ports.iter().find(|p| p.name == link_spec.output_port);
            let in_port = input.ports.iter().find(|p| p.name == link_spec.input_port);

            if let (Some(op), Some(ip)) = (out_port, in_port) {
                state.pw_handler.create_link(out.id, op.id, input.id, ip.id);
            }
        }
    }

    StatusCode::OK
}
